lib/puppet/util/windows/registry.rb

```diff
@@ -207,21 +207,30 @@ def read(key, name_ptr, *rtype)
         # buffer is raw bytes, *not* chars - less a NULL terminator
         string_length = (byte_length / FFI.type_size(:wchar)) - 1 if byte_length > 0
 
-        case type
-        when Win32::Registry::REG_SZ, Win32::Registry::REG_EXPAND_SZ
-          result = [ type, data_ptr.read_wide_string(string_length) ]
-        when Win32::Registry::REG_MULTI_SZ
-          result = [ type, data_ptr.read_wide_string(string_length).split(/\0/) ]
-        when Win32::Registry::REG_BINARY
-          result = [ type, data.read_bytes(0, byte_length) ]
-        when Win32::Registry::REG_DWORD
-          result = [ type, data_ptr.read_dword ]
-        when Win32::Registry::REG_DWORD_BIG_ENDIAN
-          result = [ type, data_ptr.order(:big).read_dword ]
-        when Win32::Registry::REG_QWORD
-          result = [ type, data_ptr.read_qword ]
-        else
-          raise TypeError, "Type #{type} is not supported."
+        begin
+          case type
+            when Win32::Registry::REG_SZ, Win32::Registry::REG_EXPAND_SZ
+              result = [ type, data_ptr.read_wide_string(string_length) ]
+            when Win32::Registry::REG_MULTI_SZ
+              result = [ type, data_ptr.read_wide_string(string_length).split(/\0/) ]
+            when Win32::Registry::REG_BINARY
+              result = [ type, data_ptr.read_bytes(byte_length) ]
+            when Win32::Registry::REG_DWORD
+              result = [ type, data_ptr.read_dword ]
+            when Win32::Registry::REG_DWORD_BIG_ENDIAN
+              result = [ type, data_ptr.order(:big).read_dword ]
+            when Win32::Registry::REG_QWORD
+              result = [ type, data_ptr.read_qword ]
+            else
+              raise TypeError, "Type #{type} is not supported."
+          end
+        rescue IndexError => ex
+          if (ex.message =~ /^Memory access .* is out of bounds$/i)
+            parent_key_name = key.parent ? "#{key.parent.keyname}\\" : ""
+            Puppet.warning "A value in the registry key #{parent_key_name}#{key.keyname} is corrupt or invalid"
+          else
+            raise ex
```

Comment:
As a minor matter of style, might want to rewrite to raise early like this:

```ruby
raise if (ex.message !~ /^Memory access .* is out of bounds$/i)
```

Note that Ruby supports re-raising to preserve original exception / stacktrace / etc, by supplying no arg to `raise` (your way is also correct).

For reference, I originally put this sample together to verify Ruby usage, as we were doing it wrong in a number of spots inside of Puppet code:

```ruby
def foo
  raise ArgumentError.new("bad arg yo")
end

def bar
  begin
    foo
  # good
  # rescue ArgumentError => e
    # raise
    # raise e, e.message, e.backtrace
    # raise e
  rescue
    # to preserve backtrace good
    # raise
    # raise $!
    # raise $!, "bad arg yo -- part 2", $!.backtrace
    # raise $!, $!.message, $!.backtrace

    #bad
    # raise $!, "Just smoked the backtrace on existing exception"
    # raise "Just smoked the backtrace and changed the exception type "
    # raise RuntimeError.new($!)
    # raise $!, $!.backtrace
  end
end

def doit
  begin
    bar
  rescue
    puts $!.class
    puts $!
    puts $!.backtrace
  end
end

doit
```